Fix traefik logs to behave like configured

marco-jantke opened the pull request:
This PR accomplishes two things: 
1. it makes sure that all traefik logs are written in the specified format. Before the first log messages were written in text, even when you specified JSON as the format. This is because some log output was produced before configuring the logs.
2. it makes sure that all traefik logs are written to the traefik logs file. This is implemented by using `logrus.Exit()` and by closing the file in a `logrus.RegisterExitHandler`.

I also changed that errors from `http.Serve` and `http.ServeTLS` are not logged anymore, in the case they are regular `http.ErrServerClosed` errors, which is expected and the normal case. Those methods will never return a non-nil error.

This fixes https://github.com/containous/traefik/issues/2156.

timoreimann commented:
Do we know why simply deferring isn't enough? Is this possibly related to a missed flush or a missing close? (I don't know if flushing stdout is even possible, just throwing around wild guesses here.)

marco-jantke commented:
Yes, I checked that deferring is not enough. This is basically the first example I have written in the description above (`# sample of log output with closing log file`). 

Frankly speaking, I also don't understand yet why this does not work. AFAICS the `run` function is executed in the main go routine and thus when it exits, everything should exit. So my understanding of go. The output that is printed to stdout, given we close the log file in a `defer` of `run` is from other goroutines like `listenSignals`. I am currently digging deeper to find the reason for this.

marco-jantke commented:
@timoreimann I was able to build a solution that captures all output by using `logrus.RegisterExitHandler`. This way we can close the files properly and still catch all the output.

After having implemented this, there was surprisingly still some output written to the stdout. Namely this: `{"level":"error","msg":"Error creating server: http: Server closed","time":"2017-09-26T09:57:15+02:00"}`. I had a look into the code and this is actually confusing in my opinion. This is the error returned from the `http.Serve` or `http.ServeTLS` which always return a non-nil error. I changed the logic in 20e4471 to output only real errors in order to avoid confusion and this way, we now have no more output on the stdout 🎉 

Updating PR description right now.

ldez commented:
Please undo the refactoring. (We can speak about this on Slack)

marco-jantke commented:
@ldez and @Juliens I changed and rebased the PR now according to our discussion. This was basically to keep the `run` method agnostic of the `traefikConfiguration`. PR is ready for review.

Juliens commented on cmd/traefik/traefik.go:
```diff
@@ -45,19 +45,17 @@ Complete documentation is available at https://traefik.io`,
+			// Try to fallback to traefik config file in case the file provider is enabled
+			// but it has no file name configured.
+			if fileProviderEnabled && len(globalConfiguration.File.Filename) == 0 && traefikConfigFileExists {
+				globalConfiguration.File.Filename = traefikConfiguration.ConfigFile
```
Can you move this code in the run func, and use the "new" config file param

marco-jantke replied:
I went one step further and pushed it down to `SetEffectiveConfiguration` as it fits there best IMO.

Juliens commented on cmd/traefik/traefik.go:
```diff
@@ -317,6 +276,64 @@ func run(globalConfiguration *configuration.GlobalConfiguration) {
+
+	if globalConfiguration.Debug {
+		globalConfiguration.LogLevel = "DEBUG"
+	}
```
this code should be in setEffectiveConfiguration ? WDYT

marco-jantke replied:
I am not sure about that. The docs of `SetEffectiveConfiguration` states:

> SetEffectiveConfiguration adds missing configuration parameters derived from existing ones. It also takes care of maintaining backwards compatibility.

In my perception, this is actually not the case.  `LogLevel` is not a missing configuration parameter. It's meant to be this way.

Additionally `SetEffectiveConfiguration` is called _after_ `configureLogging`, and that is for the good reason that the logging configuration should happen as first thing in order to achieve what this PR tries to achieve. Log all messages in the desired format and write them to a file if it is configured.

Juliens commented on cmd/traefik/traefik.go:
```diff
@@ -317,6 +276,64 @@ func run(globalConfiguration *configuration.GlobalConfiguration) {
+	logFile := globalConfiguration.TraefikLogsFile
+	if len(logFile) > 0 {
+		log.Warn("top-level traefiklogsfile has been deprecated -- please use traefiklog.filepath")
+	}
```
this code should be in setEffectiveConfiguration and change the `globalConfiguration.TraefikLog` value ? WDYT

marco-jantke replied:
I don't think so, for the same reasons stated it in https://github.com/containous/traefik/pull/2176#discussion_r142194033.

marco-jantke commented:
@Juliens I commented/addressed all your comments. PTAL again :-)

ldez commented:
```
configuration/configuration_test.go:244:32: not enough arguments in call to gc.SetEffectiveConfiguration
	have ()
	want (string)
```

marco-jantke commented:
Ups. Will fix Wednesday morning probably. Holiday :D

marco-jantke commented:
@ldez @Juliens PTAL again.

marco-jantke commented on cmd/traefik/traefik.go:
```diff
@@ -223,8 +211,23 @@ Complete documentation is available at https://traefik.io`,
 
-func run(globalConfiguration *configuration.GlobalConfiguration) {
-	fmtlog.SetFlags(fmtlog.Lshortfile | fmtlog.LstdFlags)
+func run(traefikConfiguration *TraefikConfiguration) {
```
Is this something I should address in a follow-up PR?

Also, can you explain a bit the benefit you see in this? We were discussing with @Juliens that you want to keep `run` function to not know anything about the `traefikConfiguration`, because of a future feature you were working on. Is this still aligned with this effort?
